wrappers/services.go

```diff
@@ -245,156 +185,141 @@ func StartServices(apps []*snap.AppInfo, disabledSvcs []string, flags *StartServ
-				inter.Notify(fmt.Sprintf("While trying to stop previously started service %q: %v", app.ServiceName(), e))
-			}
-			for _, socket := range app.Sockets {
-				socketService := filepath.Base(socket.File())
-				if e := sysd.Disable([]string{socketService}); e != nil {
-					inter.Notify(fmt.Sprintf("While trying to disable previously enabled socket service %q: %v", socketService, e))
-				}
-			}
-			if app.Timer != nil {
-				timerService := filepath.Base(app.Timer.File())
-				if e := sysd.Disable([]string{timerService}); e != nil {
-					inter.Notify(fmt.Sprintf("While trying to disable previously enabled timer service %q: %v", timerService, e))
-				}
-			}
-		}(app)
-
-		for _, socket := range app.Sockets {
-			socketService := filepath.Base(socket.File())
-			// enable the socket
-			if err = sysd.Enable([]string{socketService}); err != nil {
-				return err
-			}
-
-			switch app.DaemonScope {
-			case snap.SystemDaemon:
-				timings.Run(tm, "start-system-socket-service", fmt.Sprintf("start system socket service %q", socketService), func(nested timings.Measurer) {
-					err = sysd.Start([]string{socketService})
-				})
-			case snap.UserDaemon:
-				timings.Run(tm, "start-user-socket-service", fmt.Sprintf("start user socket service %q", socketService), func(nested timings.Measurer) {
-					err = startUserServices(cli, inter, socketService)
-				})
-			}
-			if err != nil {
-				return err
-			}
 		}
-
-		if app.Timer != nil {
-			timerService := filepath.Base(app.Timer.File())
-			// enable the timer
-			if err = sysd.Enable([]string{timerService}); err != nil {
-				return err
-			}
-
-			switch app.DaemonScope {
-			case snap.SystemDaemon:
-				timings.Run(tm, "start-system-timer-service", fmt.Sprintf("start system timer service %q", timerService), func(nested timings.Measurer) {
-					err = sysd.Start([]string{timerService})
-				})
-			case snap.UserDaemon:
-				timings.Run(tm, "start-user-timer-service", fmt.Sprintf("start user timer service %q", timerService), func(nested timings.Measurer) {
-					err = startUserServices(cli, inter, timerService)
-				})
-			}
-			if err != nil {
-				return err
-			}
+		if len(toEnableUser) > 0 {
+			err = userSysd.Enable(toEnableUser)
 		}
+	})
+	if err != nil {
+		return err
 	}
 
-	for _, srv := range systemServices {
-		// starting all services at once does not create a single
-		// transaction, but instead spawns multiple jobs, make sure the
-		// services started in the original order by bring them up one
-		// by one, see:
-		// https://github.com/systemd/systemd/issues/8102
-		// https://lists.freedesktop.org/archives/systemd-devel/2018-January/040152.html
-		timings.Run(tm, "start-service", fmt.Sprintf("start service %q", srv), func(nested timings.Measurer) {
-			err = systemSysd.Start([]string{srv})
-		})
-		if err != nil {
-			// cleanup was set up by iterating over apps
-			return err
+	timings.Run(tm, "start-services", "start services", func(nested timings.Measurer) {
```

Comment:
Do we need this here, since we are timing the startup of every service already?

Reply:
the timings data is marginally larger, but should provide a nice aggregate of the startup time right from the very beginning

Reply:
Heh, noticed a little bug that crept in with nested timings here. I'll push a patch

Reply:
This is what it should look like:
```
ID     Status        Doing      Undoing  Summary
24351  Done           24ms            -  Ensure prerequisites for "test-snapd-service" are available
24352  Done           16ms            -  Prepare snap "/var/lib/snapd/snaps/.local-install-3135601175" (unset)
24353  Done         1015ms            -  Mount snap "test-snapd-service" (unset)
 ^                     7ms            -    check snap "test-snapd-service"
 ^                   959ms            -    setup snap "test-snapd-service"
24354  Done            0ms            -  Copy snap "test-snapd-service" data
24355  Done          474ms            -  Setup snap "test-snapd-service" (unset) security profiles
 ^                   142ms            -    setup security backend "seccomp" for snap "test-snapd-service"
 ^                   303ms            -    setup security backend "apparmor" for 1 snaps
  ^                  177ms            -      load changed security profiles of 1 snaps
24356  Done         1012ms            -  Make snap "test-snapd-service" (unset) available to the system
 ^                   586ms            -    generate wrappers for snap test-snapd-service
 ^                   366ms            -    update font config caches
24357  Done           36ms            -  Automatically connect eligible plugs and slots of snap "test-snapd-service"
24358  Done           37ms            -  Set automatic aliases for snap "test-snapd-service"
24359  Done           32ms            -  Setup snap "test-snapd-service" aliases
24360  Done            5ms            -  Run install hook of "test-snapd-service" snap if present
24361  Done          620ms            -  Start snap "test-snapd-service" (unset) services
 ^                   417ms            -    enable services ["snap.test-snapd-service.test-snapd-sigterm-service.service" "snap.test-snapd-service.test-snapd-sigterm-all-service.service" "snap.test-snapd-service.test-snapd-sighup-all-service.service" "snap.test-snapd-service.test-snapd-sigusr1-all-service.service" "snap.test-snapd-service.test-snapd-sighup-service.service" "snap.test-snapd-service.test-snapd-service.service" "snap.test-snapd-service.test-snapd-sigusr2-service.service" "snap.test-snapd-service.test-snapd-service-refuses-to-stop.service" "snap.test-snapd-service.test-snapd-sigusr1-service.service" "snap.test-snapd-service.test-snapd-sigusr2-all-service.service" "snap.test-snapd-service.test-snapd-other-service.service" "snap.test-snapd-service.test-snapd-endure-service.service"]
 ^                   147ms            -    start services
  ^                   10ms            -      start service "snap.test-snapd-service.test-snapd-sigterm-service.service"
  ^                   11ms            -      start service "snap.test-snapd-service.test-snapd-sigterm-all-service.service"
  ^                   10ms            -      start service "snap.test-snapd-service.test-snapd-sighup-all-service.service"
  ^                    8ms            -      start service "snap.test-snapd-service.test-snapd-sigusr1-all-service.service"
  ^                   26ms            -      start service "snap.test-snapd-service.test-snapd-sighup-service.service"
  ^                    8ms            -      start service "snap.test-snapd-service.test-snapd-service.service"
  ^                    8ms            -      start service "snap.test-snapd-service.test-snapd-sigusr2-service.service"
  ^                    9ms            -      start service "snap.test-snapd-service.test-snapd-service-refuses-to-stop.service"
  ^                    8ms            -      start service "snap.test-snapd-service.test-snapd-sigusr1-service.service"
  ^                   22ms            -      start service "snap.test-snapd-service.test-snapd-sigusr2-all-service.service"
  ^                   12ms            -      start service "snap.test-snapd-service.test-snapd-other-service.service"
  ^                   10ms            -      start service "snap.test-snapd-service.test-snapd-endure-service.service"
24362  Done         3186ms            -  Run configure hook of "test-snapd-service" snap if present
24363  Done           26ms            -  Run health check of "test-snapd-service" snap
24364  Done           24ms            -  restart of [test-snapd-service.test-snapd-other-service]
24365  Done          390ms            -  Run service command "restart" for services ["test-snapd-other-service"] of snap "test-snapd-service"
 ^                   351ms            -    restart service snap.test-snapd-service.test-snapd-other-service.service
```